code/modules/chemistry/tools/robospray.dm

```diff
@@ -18,29 +17,48 @@
 	)
 	var/currentreagent = "epinephrine"
 	var/propername = "Epinephrine"
+	var/image/fluid_image
+	var/last_refill = -1
+	var/extra_refill = 0
 
 	hide_attack = 2
 	inventory_counter_enabled = 1
 
 	New()
 		..()
 		processing_items.Add(src)
+		src.update_icon()
 
 	disposing()
 		..()
 		processing_items.Remove(src)
 
+	proc/update_icon()
+		if (botreagents[currentreagent] >= 1)
+			if (!src.fluid_image)
+				src.fluid_image = image(src.icon, "hypoover", -1)
+			var/datum/reagent/R = reagents_cache[src.currentreagent]
+			src.fluid_image.color = rgb(R.fluid_r, R.fluid_g, R.fluid_b, 255)
+			src.UpdateOverlays(src.fluid_image, "fluid")
+		else
+			src.UpdateOverlays(null, "fluid")
+		src.inventory_counter.update_number(botreagents[currentreagent])
+		signal_event("icon_updated")
+
 	attack_self(mob/user as mob)
-		picker = picker % length(botreagents) + 1
-		currentreagent = botreagents[picker]
-		var/datum/reagent/temp_reagent = reagents_cache[currentreagent]
-		propername = temp_reagent.name
+		var/available_chems = list()
```

Comment:
Do we need to build this list every time? Could check if it exists and populate if not, rather than always building.